core/templates/components/button-directives/create-activity-button.component.spec.ts

```diff
@@ -0,0 +1,279 @@
+    explorationCreationService = TestBed.inject(ExplorationCreationService);
+    ngbModal = TestBed.inject(NgbModal);
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeDefined();
+  });
+
+  it('should initialize and call initCreationProcess()', fakeAsync(() => {
+    const userServiceSpy = spyOn(userService, 'getUserInfoAsync')
+      // This throws "Argument of type 'Promise<{ isLoggedIn: () => boolean;
+      // canCreateCollections: () => boolean; }>' is not assignable to parameter
+      // of type 'Promise<UserInfo>'.". This is because the actual
+      // 'getUserInfoAsync' returns more properties than required. We need to
+      // suppress this error because we need only "isLoggedIn" and
+      // "canCreateCollections" functions for testing.
+      // @ts-expect-error
+      .and.returnValue(Promise.resolve({
+        isLoggedIn: () => true,
+        canCreateCollections: () => true
+      }));
+    const urlParamsSpy = spyOn(urlService, 'getUrlParams').and.returnValue({
+      mode: 'create'
+    });
+    spyOn(component, 'initCreationProcess');
+
+    component.ngOnInit();
+    tick();
+    fixture.detectChanges();
+
+    expect(userServiceSpy).toHaveBeenCalled();
+    expect(urlParamsSpy).toHaveBeenCalled();
+    expect(component.initCreationProcess).toHaveBeenCalled();
+    expect(component.canCreateCollections).toBe(true);
+    expect(component.userIsLoggedIn).toBe(true);
+  }));
+
+  it('should initialize and call createNewExploration()', fakeAsync(() => {
+    const userServiceSpy = spyOn(userService, 'getUserInfoAsync')
+      // This throws "Argument of type 'Promise<{ isLoggedIn: () => boolean;
+      // canCreateCollections: () => boolean; }>' is not assignable to parameter
+      // of type 'Promise<UserInfo>'.". This is because the actual
+      // 'getUserInfoAsync' returns more properties than required. We need to
+      // suppress this error because we need only "isLoggedIn" and
+      // "canCreateCollections" functions for testing.
+      // @ts-expect-error
+      .and.returnValue(Promise.resolve({
+        isLoggedIn: () => true,
+        canCreateCollections: () => false
+      }));
+    const urlParamsSpy = spyOn(urlService, 'getUrlParams').and.returnValue({
+      mode: 'create'
+    });
+    const explorationCreationServiceSpy = spyOn(
+      explorationCreationService, 'createNewExploration');
+
+    component.ngOnInit();
+    tick();
+    fixture.detectChanges();
+
+    expect(userServiceSpy).toHaveBeenCalled();
+    expect(urlParamsSpy).toHaveBeenCalled();
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
+    expect(component.canCreateCollections).toBe(false);
+    expect(component.userIsLoggedIn).toBe(true);
+  }));
+
+  it('should return from initCreateProcess() when' +
+    ' creation is in progress', () => {
+    component.creationInProgress = true;
+    expect(component.initCreationProcess()).toBe();
```

Comment:
What are we expecting the output to be here? Also, do we need a pre-check?

Reply:
We are expecting it to return void. I don't think we need a pre-check here. We are just checking if the function terminates when `creationInProgress` is `true` [here](https://github.com/oppia/oppia/pull/12936/files#diff-18b57c9d65ff111b9d10b87671ac5aadd027ffeaf361d8e341ad4342c7f05f98L61-L63)

Reply:
Ah I see, thanks!

Reply:
It's not clear to me that the return value being undefined corresponds to the creation not happening. Note that it should not be necessary to refer to the actual code in order to understand the test.

Can you write this a different way? You might need to tweak some stuff in the main file to signal whether initCreationProcess() succeeded or failed. Or you might have a test that checks the side effects (e.g. if initCreationProcess() calls some creation method you can check that it is not called). In either case you should also have a "pre-check" in this test that verifies that the behaviour in question works differently when there are no other creations in progress.

Reply:
Changed the test.